lib/ansible/modules/cloud/vmware/vmware_portgroup_facts.py

```diff
@@ -98,54 +108,110 @@
 
 
 class PortgroupFactsManager(PyVmomi):
+    """Class to manage Port Group facts"""
     def __init__(self, module):
         super(PortgroupFactsManager, self).__init__(module)
         cluster_name = self.params.get('cluster_name', None)
         esxi_host_name = self.params.get('esxi_hostname', None)
         self.hosts = self.get_all_host_objs(cluster_name=cluster_name, esxi_host_name=esxi_host_name)
+        if not self.hosts:
+            self.module.fail_json(msg="Failed to find host system.")
+        self.policies = self.params.get('policies')
 
     @staticmethod
-    def normalize_pg_info(portgroup_obj):
+    def normalize_pg_info(portgroup_obj, policy_facts):
+        """Create Port Group information"""
         pg_info_dict = dict()
-        pg_info_dict['name'] = portgroup_obj.spec.name
-        vlan_id = 'N/A'
-        if portgroup_obj.spec.vlanId:
-            vlan_id = portgroup_obj.spec.vlanId
-        pg_info_dict['vlan_id'] = vlan_id
-        switch_name = 'N/A'
-        if portgroup_obj.spec.vswitchName:
-            switch_name = portgroup_obj.spec.vswitchName
-        pg_info_dict['vswitch_name'] = switch_name
-
-        # Network Policy related facts
-        pg_info_dict['promiscuous_mode'] = bool(portgroup_obj.spec.policy.security.allowPromiscuous)
-        pg_info_dict['mac_changes'] = bool(portgroup_obj.spec.policy.security.macChanges)
-        pg_info_dict['forged_transmits'] = bool(portgroup_obj.spec.policy.security.forgedTransmits)
+        spec = portgroup_obj.spec
+        pg_info_dict['portgroup'] = spec.name
+        pg_info_dict['vlan_id'] = spec.vlanId
+        pg_info_dict['vswitch'] = spec.vswitchName
+
+        if policy_facts:
+            # Security facts
+            if spec.policy.security:
+                promiscuous_mode = spec.policy.security.allowPromiscuous
+                mac_changes = spec.policy.security.macChanges
+                forged_transmits = spec.policy.security.forgedTransmits
+                pg_info_dict['security'] = (
+                    ["No override" if promiscuous_mode is None else promiscuous_mode,
+                     "No override" if mac_changes is None else mac_changes,
+                     "No override" if forged_transmits is None else forged_transmits]
+                )
+            else:
+                pg_info_dict['security'] = ["No override", "No override", "No override"]
+
+            # Traffic Shaping facts
+            if spec.policy.shapingPolicy and spec.policy.shapingPolicy.enabled is not None:
+                pg_info_dict['ts'] = portgroup_obj.spec.policy.shapingPolicy.enabled
```

Comment:
is `ts` well known word for Traffic shaping? If yes, then I am ok. Otherwise, key name can be `traffic_shapping` or something similar.

Reply:
I used `ts` just because otherwise the output is so long and traffic shaping is rarely used anyway..

Reply:
I am Ok with `ts` as short form for traffic shaping. Just add a note in documentation section stating this, so that user will be aware of this thing. Same for `lb`.

Reply:
done